compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -188,29 +239,71 @@ impl<'tcx> TypeVisitor<TyCtxt<'tcx>> for OpaqueTypeCollector<'tcx> {
             match kind {
-                // Walk over the signature of the function-like to find the opaques.
-                DefKind::AssocFn | DefKind::Fn => {
-                    let ty_sig = tcx.fn_sig(item).subst_identity();
-                    let hir_sig = tcx.hir().get_by_def_id(item).fn_sig().unwrap();
-                    // Walk over the inputs and outputs manually in order to get good spans for them.
-                    collector.visit_spanned(hir_sig.decl.output.span(), ty_sig.output());
-                    for (hir, ty) in hir_sig.decl.inputs.iter().zip(ty_sig.inputs().iter()) {
-                        collector.visit_spanned(hir.span, ty.map_bound(|x| *x));
+                DefKind::Static(_)
+                | DefKind::Const
+                | DefKind::AssocConst
+                | DefKind::AssocFn
+                | DefKind::AnonConst
+                | DefKind::InlineConst
+                | DefKind::Fn => {
+                    match kind {
+                        // Walk over the signature of the function-like to find the opaques.
+                        DefKind::AssocFn | DefKind::Fn => {
+                            let ty_sig = tcx.fn_sig(item).subst_identity();
+                            let hir_sig = tcx.hir().get_by_def_id(item).fn_sig().unwrap();
+                            // Walk over the inputs and outputs manually in order to get good spans for them.
+                            collector.visit_spanned(hir_sig.decl.output.span(), ty_sig.output());
+                            for (hir, ty) in hir_sig.decl.inputs.iter().zip(ty_sig.inputs().iter())
+                            {
+                                collector.visit_spanned(hir.span, ty.map_bound(|x| *x));
+                            }
+                        }
+                        // Walk over the type of the item to find opaques.
+                        DefKind::Static(_)
+                        | DefKind::Const
+                        | DefKind::AssocConst
+                        | DefKind::AnonConst
+                        | DefKind::InlineConst => {
+                            let span = match tcx.hir().get_by_def_id(item).ty() {
+                                Some(ty) => ty.span,
+                                _ => tcx.def_span(item),
+                            };
+                            collector.visit_spanned(span, tcx.type_of(item).subst_identity());
+                        }
+                        _ => unreachable!(),
+                    }
+                    // Look at all where bounds.
+                    tcx.predicates_of(item).instantiate_identity(tcx).visit_with(&mut collector);
+                    // An item is allowed to constrain opaques declared within its own body (but not nested within
+                    // nested functions).
+                    for id in find_taits_declared_in_body(tcx, item) {
+                        if let DefKind::TyAlias = tcx.def_kind(id) {
```

Comment:
Doesn't this mean that `find_taits_declared_in_body` lies, and just returns all nested items in the body? :laughing: 

I think this `if let` should get moved into the `fn visit_nested_item` below.